File: src/prd_tool/tracker.py
"""Local file-based issue tracker for git repos without a remote.

When a git repo has no recognised git host — no ``origin`` remote — the
prd-workflow ``forge`` selects the ``local`` provider, whose command snippets
drive this tracker instead of ``gh``/``fgj``. Issues, their labels,
dependencies, and epic milestones live in a single JSON ledger in the
consuming repo at ``docs/prd/tracker.json``. The branch workflow
(``prd/<slug>``, ``slice/<n>-<slug>``) works identically to hosted forges —
only remote operations (fetch, push, PRs) are skipped; the PRD branch merges
into ``main`` locally at finalize.

The shape deliberately mirrors what the skills already read from ``gh``/``fgj``:
issues carry a ``number``, ``title``, ``body``, ``labels``, ``state``
(``open``/``closed``), free-text ``comments``, native ``blocked_by`` dependency
edges, and an optional ``milestone`` number (the epic it belongs to). Milestones
live in a parallel ``milestones`` list — an **epic is a milestone**, matching the
git-host model; there are no epic issues or sub-issues.
"""

from __future__ import annotations

import json
from pathlib import Path


class TrackerError(Exception):
    """A local-tracker operation that can't be satisfied (e.g. unknown issue)."""


def store_path(root: Path) -> Path:
    return root / "docs" / "prd" / "tracker.json"


def _load(root: Path) -> dict:
    p = store_path(root)
    if not p.is_file():
        return {"next": 1, "issues": [], "next_milestone": 1, "milestones": []}
    data = json.loads(p.read_text(encoding="utf-8"))
    data.setdefault("next", 1)
    data.setdefault("issues", [])
    data.setdefault("next_milestone", 1)
    data.setdefault("milestones", [])
    return data


def _save(root: Path, data: dict) -> None:
    p = store_path(root)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def _find(data: dict, number: int) -> dict:
    for issue in data["issues"]:
        if issue["number"] == number:
            return issue
    raise TrackerError(f"no issue #{number} in the local tracker (docs/prd/tracker.json)")


def ensure(root: Path) -> Path:
    """Initialise the store if absent (idempotent). Returns its path."""
    _save(root, _load(root))
    return store_path(root)


def create(root: Path, title: str, body: str, labels, milestone: str | None = None) -> int:
    data = _load(root)
    number = int(data["next"])
    data["next"] = number + 1
    ms_number = _ensure_milestone(data, milestone) if milestone else None
    data["issues"].append(
        {
            "number": number,
            "title": title,
            "body": body,
            "labels": list(labels),
            "state": "open",
            "comments": [],
            "blocked_by": [],
            "milestone": ms_number,
        }
    )
    _save(root, data)
    return number


def view(root: Path, number: int) -> dict:
    return _find(_load(root), number)


def list_issues(root: Path, label: str | None = None, state: str | None = None) -> list[dict]:
    issues = _load(root)["issues"]
    if label:
        issues = [i for i in issues if label in i["labels"]]
    if state:
        issues = [i for i in issues if i["state"] == state]
    return issues


def comment(root: Path, number: int, text: str) -> None:
    data = _load(root)
    _find(data, number)["comments"].append(text)
    _save(root, data)


def close(root: Path, number: int, comment_text: str | None = None) -> None:
    data = _load(root)
    issue = _find(data, number)
    issue["state"] = "closed"
    if comment_text:
        issue["comments"].append(comment_text)
    _save(root, data)


def edit_labels(root: Path, number: int, add=(), remove=()) -> list[str]:
    data = _load(root)
    issue = _find(data, number)
    remove = set(remove)
    labels = [label for label in issue["labels"] if label not in remove]
    for label in add:
        if label not in labels:
            labels.append(label)
    issue["labels"] = labels
    _save(root, data)
    return labels


def add_dependency(root: Path, number: int, blocker: int) -> None:
    data = _load(root)
    issue = _find(data, number)
    _find(data, blocker)  # the blocker must exist
    if blocker == number:
        raise TrackerError(f"issue #{number} cannot block itself")
    if blocker not in issue["blocked_by"]:
        issue["blocked_by"].append(blocker)
    _save(root, data)


def _find_milestone(data: dict, number: int) -> dict:
    for m in data["milestones"]:
        if m["number"] == number:
            return m
    raise TrackerError(f"no milestone #{number} in the local tracker (docs/prd/tracker.json)")


def _ensure_milestone(data: dict, title: str) -> int:
    """Return the number of the milestone titled *title*, creating it if absent."""
    for m in data["milestones"]:
        if m["title"] == title:
            return m["number"]
    number = int(data["next_milestone"])
    data["next_milestone"] = number + 1
    data["milestones"].append({"number": number, "title": title, "state": "open"})
    return number


def create_milestone(root: Path, title: str) -> int:
    """Create the epic milestone (idempotent by title); returns its number."""
    data = _load(root)
    number = _ensure_milestone(data, title)
    _save(root, data)
    return number


def close_milestone(root: Path, number: int) -> None:
    data = _load(root)
    _find_milestone(data, number)["state"] = "closed"
    _save(root, data)


def list_milestones(root: Path) -> list[dict]:
    return _load(root)["milestones"]


def set_milestone(root: Path, number: int, title: str) -> int:
    """Assign issue *number* to the milestone titled *title* (created if absent)."""
    data = _load(root)
    issue = _find(data, number)
    ms_number = _ensure_milestone(data, title)
    issue["milestone"] = ms_number
    _save(root, data)
    return ms_number

File: src/core/tracker.ts
/**
 * Local file-based issue tracker for git repos without a remote.
 *
 * Issues, labels, dependencies, and epic milestones live in a single JSON ledger
 * at `docs/prd/tracker.json`. An epic is a milestone.
 */

import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { TrackerError } from "./errors.js";

export { TrackerError };

export interface Issue {
  number: number;
  title: string;
  body: string;
  labels: string[];
  state: string; // open | closed
  comments: string[];
  blocked_by: number[];
  milestone: number | null;
}

export interface Milestone {
  number: number;
  title: string;
  state: string; // open | closed
}

interface Store {
  next: number;
  issues: Issue[];
  next_milestone: number;
  milestones: Milestone[];
}

export function storePath(root: string): string {
  return join(root, "docs", "prd", "tracker.json");
}

function load(root: string): Store {
  const p = storePath(root);
  if (!existsSync(p) || !statSync(p).isFile()) {
    return { next: 1, issues: [], next_milestone: 1, milestones: [] };
  }
  const data = JSON.parse(readFileSync(p, "utf-8")) as Partial<Store>;
  return {
    next: data.next ?? 1,
    issues: data.issues ?? [],
    next_milestone: data.next_milestone ?? 1,
    milestones: data.milestones ?? [],
  };
}

function save(root: string, data: Store): void {
  const p = storePath(root);
  mkdirSync(dirname(p), { recursive: true });
  writeFileSync(p, JSON.stringify(data, null, 2) + "\n", "utf-8");
}

function find(data: Store, number: number): Issue {
  for (const issue of data.issues) if (issue.number === number) return issue;
  throw new TrackerError(`no issue #${number} in the local tracker (docs/prd/tracker.json)`);
}

function findMilestone(data: Store, number: number): Milestone {
  for (const m of data.milestones) if (m.number === number) return m;
  throw new TrackerError(`no milestone #${number} in the local tracker (docs/prd/tracker.json)`);
}

function ensureMilestoneIn(data: Store, title: string): number {
  for (const m of data.milestones) if (m.title === title) return m.number;
  const number = data.next_milestone;
  data.next_milestone = number + 1;
  data.milestones.push({ number, title, state: "open" });
  return number;
}

export function ensure(root: string): string {
  save(root, load(root));
  return storePath(root);
}

export function create(
  root: string,
  title: string,
  body: string,
  labels: Iterable<string>,
  milestone: string | null = null,
): number {
  const data = load(root);
  const number = data.next;
  data.next = number + 1;
  const msNumber = milestone ? ensureMilestoneIn(data, milestone) : null;
  data.issues.push({
    number, title, body, labels: [...labels], state: "open", comments: [],
    blocked_by: [], milestone: msNumber,
  });
  save(root, data);
  return number;
}

export function view(root: string, number: number): Issue {
  return find(load(root), number);
}

export function listIssues(root: string, label?: string | null, state?: string | null): Issue[] {
  let issues = load(root).issues;
  if (label) issues = issues.filter(i => i.labels.includes(label));
  if (state) issues = issues.filter(i => i.state === state);
  return issues;
}

export function comment(root: string, number: number, text: string): void {
  const data = load(root);
  find(data, number).comments.push(text);
  save(root, data);
}

export function close(root: string, number: number, commentText?: string | null): void {
  const data = load(root);
  find(data, number).state = "closed";
  if (commentText) find(data, number).comments.push(commentText);
  save(root, data);
}

export function edit(root: string, number: number, title?: string | null, body?: string | null): void {
  const data = load(root);
  const issue = find(data, number);
  if (title !== null && title !== undefined) issue.title = title;
  if (body !== null && body !== undefined) issue.body = body;
  save(root, data);
}

export function editLabels(root: string, number: number, add: Iterable<string> = [], remove: Iterable<string> = []): string[] {
  const data = load(root);
  const issue = find(data, number);
  const removeSet = new Set(remove);
  const labels = issue.labels.filter(l => !removeSet.has(l));
  for (const label of add) if (!labels.includes(label)) labels.push(label);
  issue.labels = labels;
  save(root, data);
  return labels;
}

export function addDependency(root: string, number: number, blocker: number): void {
  const data = load(root);
  const issue = find(data, number);
  find(data, blocker);
  if (blocker === number) throw new TrackerError(`issue #${number} cannot block itself`);
  if (!issue.blocked_by.includes(blocker)) issue.blocked_by.push(blocker);
  save(root, data);
}

export function createMilestone(root: string, title: string): number {
  const data = load(root);
  const number = ensureMilestoneIn(data, title);
  save(root, data);
  return number;
}

export function closeMilestone(root: string, number: number): void {
  const data = load(root);
  findMilestone(data, number).state = "closed";
  save(root, data);
}

export function listMilestones(root: string): Milestone[] {
  return load(root).milestones;
}

export function setMilestone(root: string, number: number, title: string): number {
  const data = load(root);
  const issue = find(data, number);
  const msNumber = ensureMilestoneIn(data, title);
  issue.milestone = msNumber;
  save(root, data);
  return msNumber;
}